src/cross_section/fillet2d.cpp

```diff
@@ -0,0 +1,946 @@
+    const bool e1CCW,  // p1 -> p2 CCW ?
+    const vec2& p3, const vec2& p4,
+    const bool e2CCW,  // p3 -> p4 CCW ?
+    const double radius) {
+  // Two boundary shouldn't touch, this should be handle previous
+  if (e1CCW && e2CCW) throw std::exception();
+
+  const vec2 e1 = p2 - p1, e2 = p4 - p3,
+             normal1 =
+                 la::normalize(e1CCW ? vec2(-e1.y, e1.x) : vec2(e1.y, -e1.x));
+
+  const double det = la::cross(e1, e2);
+  const bool e2Parallel = std::abs(det) < EPSILON;
+
+  if (e2Parallel) {
+    double distance = la::dot(p3 - p1, normal1);
+
+    if (distance >= radius) return IntersectStadiumResult::Outside;
+  }
+
+  // Check is intersect at endpoint, result in degenerate
+  {
+    const bool p1Intersect =
+        intersectCircleSegment(p3, p4, p1 + normal1 * radius, radius);
+
+    const bool p2Intersect =
+        intersectCircleSegment(p3, p4, p2 + normal1 * radius, radius);
+
+    if (p1Intersect && p2Intersect)
+      return IntersectStadiumResult::P1P2Degenerate;
+
+    if (p1Intersect) {
+      bool sign = la::cross(e1, e2) > 0;
+
+      if (sign == (e1CCW ^ e2CCW)) {
+        // TODO: Ensure degenerate processed
+        return IntersectStadiumResult::P1Degenerate;
+      } else {
+        return IntersectStadiumResult::EdgeEdgeIntersect;
+      }
+    }
+
+    if (p2Intersect) {
+      bool sign = la::cross(e1, e2) < 0;
+
+      if (sign == (e1CCW ^ e2CCW)) {
+        // Degenerate process
+        return IntersectStadiumResult::P2Degenerate;
+      } else {
+        return IntersectStadiumResult::EdgeEdgeIntersect;
+      }
+    }
+  }
+
+  // Check edge intersect
+  {
+    if (intersectSegmentSegment(p1, p2, p3, p4))
+      return IntersectStadiumResult::EdgeEdgeIntersect;
+  }
+
+  // Now e2 is either both in or both out, check this
+  {
+    const vec2 p1Offset = p1 + normal1 * 2.0 * radius,
+               p2Offset = p2 + normal1 * 2.0 * radius;
+
+    // Determine full inside or full outside
+    auto isInsideRect = [&](const vec2 p) -> bool {
+      int sign = e1CCW ? 1 : -1;
+      if ((sign * la::cross(e1, p - p1) >= 0) &&
+          (sign * la::cross(p2Offset - p2, p - p2) >= 0) &&
+          (sign * la::cross(p1Offset - p2Offset, p - p2Offset) >= 0) &&
+          (sign * la::cross(p1 - p1Offset, p - p1Offset) >= 0))
+        return true;
+    };
+
+    bool p3InsideRect = isInsideRect(p3), p4InsideRect = isInsideRect(p4);
+    if (p3InsideRect && p4InsideRect) {
+      // Full inside, no intersect with boundary
+      // FIXME: both inside, which must lead to EdgeEdgeIntersect, but
+      // rethink about it.
+      if (e2Parallel) return IntersectStadiumResult::E2Degenerate;
+
+      return IntersectStadiumResult::EdgeEdgeIntersect;
+    } else if (!p3InsideRect && !p4InsideRect) {
+      // Full outside, no intersect with boundary
+      return IntersectStadiumResult::Outside;
+    } else {
+      // Should be both in or both out
+      throw std::exception();
```

Comment:
We wrap any exceptions in `#ifdef MANIFOLD_DEBUG` - search e.g. polygon.cpp for `throw` and you'll see. You might try one of our ASSERT methods instead.